### PR TITLE
Align DatetimeIndex APIs with pandas 2.x

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
Mostly transferring methods that were defined on `Series.dt` methods to `DatetimeColumn` so it could be reused in `DatetimeIndex`

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -1831,6 +1832,210 @@ def searchsorted(
...
+        Return an ndarray of ``datetime.datetime`` objects.
+
+        Returns
+        -------
+        numpy.ndarray
+            An ndarray of ``datetime.datetime`` objects.
+        """
+        return self.to_pandas().to_pydatetime()
+
+    def to_julian_date(self) -> Index:
```
Comment (galipremsagar) [inline python/cudf/cudf/core/index.py:1932]: Can we add pytests to cover these newly added APIs?

Comment (mroeschke) [reply python/cudf/cudf/core/index.py:1932]: Sure added tests in the latest commit

File: python/cudf/cudf/tests/test_datetime.py [hunk 2]
```diff
@@ -2429,3 +2430,101 @@ def test_day_month_name_locale_not_implemented(meth, klass):
...
+def test_dti_reduction(method, kwargs, request):
+    pd_dti = pd.DatetimeIndex(["2020-01-01", "2020-12-31"], name="foo")
+    cudf_dti = cudf.from_pandas(pd_dti)
+
+    result = getattr(cudf_dti, method)(**kwargs)
+    expected = getattr(pd_dti, method)(**kwargs)
+    if method == "std":
+        request.applymarker(
+            pytest.mark.xfail(reason="cudf returns imprecise nanoseconds")
+        )
```
Comment (wence-) [inline python/cudf/cudf/tests/test_datetime.py:2512]: *nit* move this up to the parametrization as:

```
pytest.param("std", marks=pytest.mark.xfail(...))
```

Also, do we have an open issue for this? No, but opened https://github.com/rapidsai/cudf/issues/16444

**Review by wence- — APPROVED** [review]
Tiny nit

**Review by galipremsagar — APPROVED** [review]

**Pull request merged**